### PR TITLE
CI improvements, Inference improvements for recursive functions, Downstream testing

### LINKED ISSUE #80: ResumableFunctions v0.6.6 StackOverflow with julia-1.10
My package https://github.com/marcom/FoldRNA.jl/ uses ResumableFunctions for the function `allstruct`.

https://github.com/marcom/FoldRNA.jl/blob/main/src/allstruct.jl

On julia-1.10 and with ResumableFunctions 0.6.6, I get a StackOverflow with a neverending(?) stacktrace from julia internals. Only `kill -9` seems to stop the process in reasonable time.

With ResumableFunctions v0.6.5 everything works as expected, so I am releasing a new version of FoldRNA now keeping ResumableFunctions at v0.6.5 in compat.

To reproduce:
```
# ensure ResumableFunctions compat is set to =0.6.6
using FoldRNA
collect(allstruct(10))
```

<details>
<summary>Output for julia-1.10, ResumableFunctions v0.6.6</summary>

```
Internal error: encountered unexpected error during compilation of show:
StackOverflowError()
emit_expr at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/codegen.cpp:5421
emit_call at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/codegen.cpp:4569
emit_expr at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/codegen.cpp:5511
emit_ssaval_assign at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/codegen.cpp:5101
emit_stmtpos at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/codegen.cpp:5335 [inlined]
emit_function at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/codegen.cpp:8380
jl_emit_code at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/codegen.cpp:8715
jl_emit_codeinst at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/codegen.cpp:8789
_jl_compile_codeinst at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/jitlayers.cpp:221
jl_generate_fptr_impl at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/jitlayers.cpp:518
jl_compile_method_internal at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:2480 [inlined]
jl_compile_method_internal at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:2368
_jl_invoke at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:2886 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:3076
print at ./boot.jl:584
jfptr_print_25803.1 at /home/mcm/.julia/juliaup/julia-1.10.0+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:2894 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:3076
println at ./boot.jl:587
println at ./boot.jl:591
jfptr_println_25813.1 at /home/mcm/.julia/juliaup/julia-1.10.0+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:2894 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:3076
fsmi_generator at /home/mcm/.julia/packages/ResumableFunctions/kJXjK/src/utils.jl:153
unknown function (ip: 0x7f905211789c)
_jl_invoke at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:2894 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:3076
jl_call_staged at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/method.c:540
ijl_code_for_staged at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/method.c:593
get_staged at ./compiler/utilities.jl:123
retrieve_code_info at ./compiler/utilities.jl:135 [inlined]
InferenceState at ./compiler/inferencestate.jl:430
typeinf_edge at ./compiler/typeinfer.jl:920
abstract_call_method at ./compiler/abstractinterpretation.jl:629
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:95
abstract_call_known at ./compiler/abstractinterpretation.jl:2083
abstract_call at ./compiler/abstractinterpretation.jl:2165
abstract_call at ./compiler/abstractinterpretation.jl:2158
abstract_call at ./compiler/abstractinterpretation.jl:2350
abstract_eval_call at ./compiler/abstractinterpretation.jl:2366
abstract_eval_statement_expr at ./compiler/abstractinterpretation.jl:2376
abstract_eval_statement at ./compiler/abstractinterpretation.jl:2620
abstract_eval_basic_statement at ./compiler/abstractinterpretation.jl:2885
typeinf_local at ./compiler/abstractinterpretation.jl:3094
typeinf_nocycle at ./compiler/abstractinterpretation.jl:3182
_typeinf at ./compiler/typeinfer.jl:247
typeinf at ./compiler/typeinfer.jl:216
typeinf_edge at ./compiler/typeinfer.jl:930
abstract_call_method at ./compiler/abstractinterpretation.jl:629
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:95
abstract_call_known at ./compiler/abstractinterpretation.jl:2083
abstract_call at ./compiler/abstractinterpretation.jl:2165
abstract_call at ./compiler/abstractinterpretation.jl:2158
abstract_call at ./compiler/abstractinterpretation.jl:2350
abstract_eval_call at ./compiler/abstractinterpretation.jl:2366
abstract_eval_statement_expr at ./compiler/abstractinterpretation.jl:2376
abstract_eval_statement at ./compiler/abstractinterpretation.jl:2620
abstract_eval_basic_statement at ./compiler/abstractinterpretation.jl:2885
typeinf_local at ./compiler/abstractinterpretation.jl:3094
typeinf_nocycle at ./compiler/abstractinterpretation.jl:3182
_typeinf at ./compiler/typeinfer.jl:247
typeinf at ./compiler/typeinfer.jl:216
typeinf_frame at ./compiler/typeinfer.jl:995
#code_typed_by_type#11 at /home/mcm/.julia/packages/ResumableFunctions/kJXjK/src/utils.jl:139
code_typed_by_type at /home/mcm/.julia/packages/ResumableFunctions/kJXjK/src/utils.jl:131 [inlined]
fsmi_generator at /home/mcm/.julia/packages/ResumableFunctions/kJXjK/src/utils.jl:151
unknown function (ip: 0x7f905211789c)
_jl_invoke at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:2894 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/gf.c:3076
jl_call_staged at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/method.c:540
ijl_code_for_staged at /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/src/method.c:593
get_staged at ./compiler/utilities.jl:123
retrieve_code_info at ./compiler/utilities.jl:135 [inlined]
InferenceState at ./compiler/inferencestate.jl:430
typeinf_edge at ./compiler/typeinfer.jl:920
abstract_call_method at ./compiler/abstractinterpretation.jl:629
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:95
abstract_call_known at ./compiler/abstractinterpretation.jl:2083
abstract_call at ./compiler/abstractinterpretation.jl:2165
abstract_call at ./compiler/abstractinterpretation.jl:2158
abstract_call at ./compiler/abstractinterpretation.jl:2350
abstract_eval_call at ./compiler/abstractinterpretation.jl:2366
abstract_eval_statement_expr at ./compiler/abstractinterpretation.jl:2376
[... continues seemingly endlessly ...]
```

</details>

### LINKED ISSUE #82: tests for downstream breakage
We should probably start running tests against downstream packages given #80 

Here is a typical github action config for downstream tests https://github.com/qojulia/QuantumInterface.jl/blob/main/.github/workflows/breakage.yml

An incomplete list (please edit and add) of packages that can be added

- [ ] ConcurrentSim (the initial reason for the creation of ResumableFunctions)
- [ ] QuantumSavory (the reason I am involved with ResumableFunctions -- a very heavy package though, so having a subset of its tests is something I need to look into)
- [ ] FoldRNA (where issue #80 was detected)
- [ ] Fronts (uses delegation via `@yieldfrom`)

### LINKED ISSUE #86: 0.6.7 breaks resumables with multiple fors over same variable
From version 0.6.7, the code below fails with `ERROR: LoadError: MethodError: Cannot `convert` an object of type String to an object of type Nothing`. It works fine with 0.6.6.

This happens because the same variable is used for multiple loops, and the type of that variable/slot in the generated iterator is determined by the last loop and only through code analysis, not the signature of `_empty`. This means that `s` is types as `Nothing`, causing the mentioned error in the first loop on actual assignment to `s`.

```
using ResumableFunctions

@resumable function singleton()::String
    @yield "anything"
end

# 1. must be defined before combined!
@resumable function _empty(x)::String
end

@resumable function combined()::String
    for s in singleton()
        @yield s # fails "here" because s is determined to be of type Nothing because of the second loop
    end
    # 2. must reuse of variable
    for s in _empty(1) # dummy argument needed to illustrate point 1.
        @yield s
    end
end

collect(combined())
```

The same happens with the (undocumented) `@yieldfrom`:
```
using ResumableFunctions

@resumable function singleton()::String
    @yield "anything"
end

@resumable function _empty()::String
end

@resumable function combined()::String
    # @yieldfrom also uses the same variable names for each generated loop
    @yieldfrom singleton()
    @yieldfrom _empty()
end

collect(combined())
```

### PR DESCRIPTION
**Pull request by Krastanov**
Fixes #86 
Fixes #82 
Fixes #80  (actual problem was recursive functions)

### CODE REVIEW

**Comment by Krastanov** [comment]
@gerlero , I reapplied the improvements from 0.6.6, added donwnstream testing, and improved the CI a bit. Could you look through it -- should be fairly minor changes.

@pepijndevos , the changes you made to improve inference seem to cause a crash in the compiler when recursive resumable functions are employed. There is probably an elegant way to solve this, but for the moment I simply disabled your improvements when a recursive function is detected. **Could you skim through [the commit that makes that change](https://github.com/JuliaDynamics/ResumableFunctions.jl/pull/88/commits/c039a4edd06c70c03dc8e2be51442f0d73eba96a)?** Please do not hesitate to file issues with ideas of how to do this in a better way.

Here are the [newly added tests specifically for recursive functions](https://github.com/JuliaDynamics/ResumableFunctions.jl/pull/88/commits/30cd94a3ff9f841d445441b02303325047983269#diff-7d4d12197dde056c2a8516eceb821f0931087329114bc903c527931e015b2c9f)

The commits are meant to be read separately, not squashed on merge. Hopefully that would make reviewing much easier as the vast majority of added lines are simply for CI configuration.

**Comment by pepijndevos** [comment]
I think this is a reasonable precaution. If you think through what the generic struct would look like for a recursive function it's an infinite recursion at the type level, and the way we do it also results in basically a function with infinite arguments.

**Comment by gerlero** [comment]
> @gerlero , I reapplied the improvements from 0.6.6, added donwnstream testing, and improved the CI a bit. Could you look through it -- should be fairly minor changes.

@Krastanov Sure! I'll review it in the next few days.

File: test/test_repeated_variable.jl [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    for s in singleton()
+        @yield s # fails "here" because s is determined to be of type Nothing because of the second loop
+    end
+    # 2. must reuse of variable
+    for s in _empty(1) # dummy argument needed to illustrate point 1.
+        @yield s
+    end
+end
+
+collect(combined())
```
Comment (gerlero) [inline test/test_repeated_variable.jl:26]: Maybe we could `@test` the return value of `collect` here so it gets counted as a test?

File: test/test_repeated_variable.jl [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+@resumable function _emptyF()::String
+end
+
+@resumable function combinedF()::String
+    # @yieldfrom also uses the same variable names for each generated loop
+    @yieldfrom singletonF()
+    @yieldfrom _emptyF()
+end
+
+collect(combinedF())
```
Comment (gerlero) [inline test/test_repeated_variable.jl:43]: Same

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,9 +1,5 @@
 # News
 
-## v0.6.7 - 2024-01-02
-
-- Fix stack overflow errors by reverting the changes introduced in v0.6.6.
-
```
Comment (gerlero) [inline CHANGELOG.md]: Why are we altering the changelog here?

Comment (gerlero) [reply CHANGELOG.md]: Oh, never mind. I see the next commit adds it back.

File: src/macro.jl
Comment (gerlero) [inline src/macro.jl:1]: Skipping this entire commit from my review as I guess I can safely assume it's a straight cherry pick of #76

**Review by gerlero — APPROVED** [review]
Looks good to me. I'd add that couple of `@test` macros, but AFAIC this can be merged.

File: src/macro.jl
Comment (Krastanov) [reply src/macro.jl:1]: indeed

File: .github/workflows/breakage.yml [hunk 4]
```diff
@@ -18,9 +18,8 @@ jobs:
           "JuliaDynamics/ConcurrentSim.jl",
           "gerlero/Fronts.jl",
           "QuantumSavory/QuantumSavory.jl",
-          "Tortar/IteratorSampling.jl",
           "marcom/FoldRNA.jl",
-          "aclai-lab/ModalDecisionTrees.jl",
```
[the remark this answers is not in the document]
Comment (Krastanov) [reply .github/workflows/breakage.yml]: yes, this one does not have a passing test suite independently of ResumableFunctions so I removed it

**Pull request merged**